c10/core/impl/InlineDeviceGuard.h

```diff
@@ -99,7 +99,9 @@ class InlineDeviceGuard {
   InlineDeviceGuard& operator=(InlineDeviceGuard<T>&& other) = delete;
 
   ~InlineDeviceGuard() {
-    impl_.uncheckedSetDevice(original_device_);
+    if (current_device_.index() != original_device_.index()) {
```

Comment:
I believe you can (should?) use == to compare the current_device and original_device, see https://github.com/pytorch/pytorch/blob/9b4bbaab5302897b57e2d87e6956824c7a61c603/c10/core/Device.h#L49

Reply:
There are some funny interactions somewhere that makes the test https://github.com/pytorch/pytorch/blob/master/c10/test/core/impl/InlineDeviceGuard_test.cpp fail if I am comparing `Device`, I don't know why and I don't know whether changing it to comparing index would fix it.

Reply:
```
Mar 26 23:44:13 ++ basename /var/lib/jenkins/.local/lib/python3.6/site-packages/torch/test/c10_InlineDeviceGuard_test
Mar 26 23:44:13 + LD_LIBRARY_PATH=/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/lib
Mar 26 23:44:13 + /var/lib/jenkins/.local/lib/python3.6/site-packages/torch/test/c10_InlineDeviceGuard_test --gtest_output=xml:/var/lib/jenkins/workspace/caffe2_tests/cpp/c10_InlineDeviceGuard_test.xml
Mar 26 23:44:13 Running main() from /var/lib/jenkins/workspace/third_party/googletest/googletest/src/gtest_main.cc
Mar 26 23:44:13 [==========] Running 9 tests from 2 test cases.
Mar 26 23:44:13 [----------] Global test environment set-up.
Mar 26 23:44:13 [----------] 5 tests from InlineDeviceGuard
Mar 26 23:44:13 [ RUN      ] InlineDeviceGuard.Constructor
Mar 26 23:44:13 /var/lib/jenkins/workspace/c10/test/core/impl/InlineDeviceGuard_test.cpp:36: Failure
Mar 26 23:44:13 Expected equality of these values:
Mar 26 23:44:13   TestGuardImpl::getDeviceIndex()
Mar 26 23:44:13     Which is: 4
Mar 26 23:44:13   init_i
Mar 26 23:44:13     Which is: 0
Mar 26 23:44:13 [  FAILED  ] InlineDeviceGuard.Constructor (0 ms)
Mar 26 23:44:13 [ RUN      ] InlineDeviceGuard.ConstructorError
Mar 26 23:44:13 [       OK ] InlineDeviceGuard.ConstructorError (0 ms)
Mar 26 23:44:13 [ RUN      ] InlineDeviceGuard.SetDevice
Mar 26 23:44:13 [       OK ] InlineDeviceGuard.SetDevice (0 ms)
Mar 26 23:44:13 [ RUN      ] InlineDeviceGuard.ResetDevice
Mar 26 23:44:13 [       OK ] InlineDeviceGuard.ResetDevice (0 ms)
Mar 26 23:44:13 [ RUN      ] InlineDeviceGuard.SetIndex
Mar 26 23:44:13 [       OK ] InlineDeviceGuard.SetIndex (0 ms)
Mar 26 23:44:13 [----------] 5 tests from InlineDeviceGuard (0 ms total)
Mar 26 23:44:13 
Mar 26 23:44:13 [----------] 4 tests from InlineOptionalDeviceGuard
Mar 26 23:44:13 [ RUN      ] InlineOptionalDeviceGuard.Constructor
Mar 26 23:44:13 /var/lib/jenkins/workspace/c10/test/core/impl/InlineDeviceGuard_test.cpp:127: Failure
Mar 26 23:44:13 Expected equality of these values:
Mar 26 23:44:13   TestGuardImpl::getDeviceIndex()
Mar 26 23:44:13     Which is: 4
Mar 26 23:44:13   init_i
Mar 26 23:44:13     Which is: 0
Mar 26 23:44:13 [  FAILED  ] InlineOptionalDeviceGuard.Constructor (0 ms)
Mar 26 23:44:13 [ RUN      ] InlineOptionalDeviceGuard.NullaryConstructor
Mar 26 23:44:13 [       OK ] InlineOptionalDeviceGuard.NullaryConstructor (0 ms)
Mar 26 23:44:13 [ RUN      ] InlineOptionalDeviceGuard.SetDevice
Mar 26 23:44:13 [       OK ] InlineOptionalDeviceGuard.SetDevice (0 ms)
Mar 26 23:44:13 [ RUN      ] InlineOptionalDeviceGuard.SetIndex
Mar 26 23:44:13 [       OK ] InlineOptionalDeviceGuard.SetIndex (0 ms)
Mar 26 23:44:13 [----------] 4 tests from InlineOptionalDeviceGuard (0 ms total)
Mar 26 23:44:13 
Mar 26 23:44:13 [----------] Global test environment tear-down
Mar 26 23:44:13 [==========] 9 tests from 2 test cases ran. (0 ms total)
Mar 26 23:44:13 [  PASSED  ] 7 tests.
Mar 26 23:44:13 [  FAILED  ] 2 tests, listed below:
Mar 26 23:44:13 [  FAILED  ] InlineDeviceGuard.Constructor
Mar 26 23:44:13 [  FAILED  ] InlineOptionalDeviceGuard.Constructor
Mar 26 23:44:13 
Mar 26 23:44:13  2 FAILED TESTS
```

Reply:
https://github.com/pytorch/pytorch/blob/master/c10/test/core/impl/InlineDeviceGuard_test.cpp#L29
Why are we secretly setting the device in `TestGuardImpl` without letting `InlineDeviceGuard` know? Is this expected behavior?

Reply:
I wonder if this test is verifying the contract that when the device guard exits it always resets the device. This seems reasonable to me, and also like your change would invalidate that contract?

@ngimel for a second opinion

Reply:
This implementation will invalidate that contract (if there is actually such a contract). But we can use your other suggestion `maybe add a conditional to uncheckedSetDevice` should not have this issue.